openedx_tagging/core/tagging/models.py

```diff
@@ -1,39 +1,440 @@
 """ Tagging app data models """
+from typing import List, Type
+
 from django.db import models
+from django.utils.translation import gettext_lazy as _
+
+# Maximum depth allowed for a hierarchical taxonomy's tree of tags.
+TAXONOMY_MAX_DEPTH = 3
+
+# Ancestry of a given tag; the Tag.value fields of a given tag and its parents, starting from the root.
+# Will contain TAXONOMY_MAX_DEPTH elements.
+Lineage = List[str]
 
 
-class TagContent(models.Model):
+class Tag(models.Model):
+    """
+    Represents a single value in a list or tree of values which can be applied to a particular Open edX object.
+
+    Open edX tags are "name:value" pairs which can be applied to objects like content libraries, units, or people.
+    Tag.taxonomy.name provides the "name" and the Tag.value provides the "value".
+    (And an ObjectTag links a Tag with an object.)
     """
-    TODO: This is a basic representation of TagContent, it may change
-    depending on the discovery in:
-    https://docs.google.com/document/d/13zfsGDfomSTCp_G-_ncevQHAb4Y4UW0d_6N8R2PdHlA/edit#heading=h.o6fm1hktwp7b
 
-    This is the representation of a tag that is linked to a content
+    id = models.BigAutoField(primary_key=True)
+    taxonomy = models.ForeignKey(
+        "Taxonomy",
+        null=True,
+        default=None,
+        on_delete=models.SET_NULL,
+        help_text=_("Namespace and rules for using a given set of tags."),
+    )
+    parent = models.ForeignKey(
+        "self",
+        null=True,
+        default=None,
+        on_delete=models.SET_NULL,
+        related_name="children",
+        help_text=_(
+            "Tag that lives one level up from the current tag, forming a hierarchy."
+        ),
+    )
+    value = models.TextField(
+        help_text=_(
+            "Content of a given tag, occupying the 'value' part of the key:value pair."
+        ),
+    )
+    external_id = models.CharField(
+        max_length=255,
+        null=True,
+        blank=True,
+        help_text=_(
+            "Used to link an Open edX Tag with a tag in an externally-defined taxonomy."
+        ),
+    )
+
+    class Meta:
+        unique_together = ["taxonomy", "parent_id", "value"]
```

Comment:
I don't think this unique constraint will work as expected when `parent_id` is NULL, because null values are always different from each other. So you can have multiple entries with the same value and a NULL parent in the same taxonomy.

I think you can achieve what you want using [`UniqueConstraint`](https://docs.djangoproject.com/en/4.2/ref/models/constraints/#uniqueconstraint) which is more flexible.